flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesLeaderElector.java

```diff
@@ -86,12 +117,38 @@ public KubernetesLeaderElector(
                                                         newLeader,
                                                         leaderConfig.getConfigMapName())))
                         .build();
+        this.executorService = executorService;
+
+        LOG.info(
+                "Create KubernetesLeaderElector on lock {}.",
+                leaderElectionConfig.getLock().describe());
+    }
+
+    @GuardedBy("lock")
+    private void resetInternalLeaderElector() {
+        stopLeaderElectionCycle();
+
         internalLeaderElector =
                 new LeaderElector(kubernetesClient, leaderElectionConfig, executorService);
+        currentLeaderElectionSession = internalLeaderElector.start();
+
         LOG.info(
-                "Create KubernetesLeaderElector {} with lock identity {}.",
-                leaderConfig.getConfigMapName(),
-                leaderConfig.getLockIdentity());
+                "Triggered leader election on lock {}.", leaderElectionConfig.getLock().describe());
+    }
+
+    @GuardedBy("lock")
+    private void stopLeaderElectionCycle() {
+        if (internalLeaderElector != null) {
+            Preconditions.checkNotNull(currentLeaderElectionSession);
+
+            // the current leader election cycle needs to be cancelled before releasing the lock to
+            // avoid retrying
+            currentLeaderElectionSession.cancel(true);
+            currentLeaderElectionSession = null;
+
+            internalLeaderElector.release();
```

Comment:
Executing the `LeaderElector#release()` will trigger another `notLeader()` callback, which revoke the leader contender leadership again. It might be harmless because the leader contender already lost leadership.

Could you please share me why we do not simply add `.withReleaseOnCancel(true)` when building the `leaderElectionConfig`?

Reply:
How will the `LeaderElector#release()` call trigger another `notLeader()` callback?

`KubernetesLeaderElector#stopLeaderElectionCycle` is called in two places:
* `KubernetesLeaderElector#stop`
  * State `leadership acquired`: The renew cycle of the `internalLeaderElector` will be cancelled which triggers the `LeaderCallbackHandler#notLeader` callback in [LeaderElector:96](https://github.com/fabric8io/kubernetes-client/blob/0f6c696509935a6a86fdb4620caea023d8e680f1/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/extended/leaderelection/LeaderElector.java#L96). The v6.6.2 implementation of `LeaderElector#stopLeading` either calls release (if we enable `ReleaseOnCall`) or calls the `notLeader` callback. The bugfix for this is https://github.com/fabric8io/kubernetes-client/commit/0f6c6965 and ended in v6.9.0. Without upgrading the k8s client dependency, we have to call release explicitly to trigger the cleanup.
  * State `leadership lost` (assuming that a `KubernetesLeaderElector#run` happened and the elector is initialized): The `#release` call won't have any affect because of the if condition in [LeaderElector:136](https://github.com/fabric8io/kubernetes-client/blob/v6.6.2/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/extended/leaderelection/LeaderElector.java#L136).
* `KubernetesLeaderElector#resetInternalLeaderElector` which is called as part of the `KubernetesLeaderElector#run` method
  * State `leadership acquired`: That case shouldn't happen because we only call `#run` after the leadership is lost. If we would do it anyway, it would trigger the cancellation of leadership analogously to what happens for the `KubernetesLeaderElectorl#stop` call with leadership acquired described above. A new leaderelection lifecycle will be initiated afterwards in `KubernetesLeaderElector#resetInternalLeaderElector` call.
  * State `leadership lost`: The `KubernetesLeaderElector#currentLeaderElectionSession` is already completed. Therefore, cancelling this future doesn't have any effect. The subsequent `LeaderElector#release` call won't have any effect as well because of the if condition in [LeaderElector:136](https://github.com/fabric8io/kubernetes-client/blob/v6.6.2/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/extended/leaderelection/LeaderElector.java#L136) again.

I created FLINK-34243 and will add a comment to cover this

Reply:
What I mean is `release()` -> `updateObserved(newLeaderElectionRecord)` -> `leaderElectionConfig.getLeaderCallbacks().onStopLeading();` will also trigger the `notLeader()`. Right?

Reply:
Oh true, I should have looked for other calls of the `notLeader` method. Thanks for your clarification. I'm gonna check. :+1:

Reply:
BTW, the ITCase could also pass by using v6.6.2 with `.withReleaseOnCancel(true)` and removing `ReleaseOnCall`.

Reply:
You're right. The release call would trigger the `notLeader` call when enabling `ReleaseOnCancel`. That was the wrong conclusion on my side. 

But considering that there's this bug in https://github.com/fabric8io/kubernetes-client/issues/5463; should we consider upgrading the client for Flink 1.18? I tend to lean towards not upgrading the kubernetes-client. The diff between v6.6.2 and v6.8.0 appears to be quite big. :thinking:

Reply:
I verified with the changed test that your finding is correct:
```diff
diff --git a/flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesLeaderElectorITCase.java b/flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesLeaderElectorITCase.java
index 51631dc54f6..0bca6c4d641 100644
--- a/flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesLeaderElectorITCase.java
+++ b/flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesLeaderElectorITCase.java
@@ -182,6 +182,9 @@ class KubernetesLeaderElectorITCase {
             // revoking the leadership initiates another leadership lifecycle
             testInstance.run();
 
+            // leadership should be lost eventually due to the renewal loop being stopped
+            leadershipCallbackHandler.waitForRevokeLeader();
+
             // triggers acquiring the leadership again
             executorService.trigger(waitForNextTaskForever);
```

I didn't address this in a test because it's hard to come up with a condition that checks the non-existence of an event (either we would have to add a timeout which might make the test unstable or the test would take longer than necessary). WDYT?

Reply:
I agree with you that we do not need to add `leadershipCallbackHandler.waitForRevokeLeader();` to the ITCase.

Reply:
If we enable the `isReleaseOnCancel` and do not upgrade the fabricio k8s client from v6.6.2 to v6.9.0, which include the fix https://github.com/fabric8io/kubernetes-client/issues/5463, then we have the risk that the `onStopLeading()` is never executed though the leadership lost. From the Flink's perspective, the `revokeLeadership` is never called while a new `grantLeadership` happens.

Reply:
Reconsidering what I said before - upgrading even for 1.18 might be necessary. We don't want to keep another (known) bug lingering around in Flink 1.18.

I upgraded the dependency to 6.9.0 and will do a pass through the diff between v6.6.2 and v6.9.0 to check whether there's something else we have to keep in mind. :thinking: